wandb/integration/kfp/kfp_patch.py

```diff
@@ -0,0 +1,212 @@
+import inspect
+import itertools
+import textwrap
+from typing import Callable, List, Mapping, Optional
+
+from kfp.components import structures
+from kfp.components._components import _create_task_factory_from_component_spec
+from kfp.components._python_op import _func_to_component_spec
+import wandb
+
+from .wandb_logging import wandb_log
+
+decorator_code = inspect.getsource(wandb_log)
+wandb_logging_extras = f"""
+import typing
+from typing import NamedTuple
+
+import collections
+from collections import namedtuple
+
+import kfp
+from kfp import components
+from kfp.components import InputPath, OutputPath
+
+import wandb
+
+{decorator_code}
+"""
+
+
+def patch(func, module):
+    wandb.patched.setdefault(module.__name__, [])
+    setattr(module, f"orig_{func.__name__}", getattr(module, func.__name__))
+    setattr(module, func.__name__, func)
+    wandb.patched[module.__name__].append([module, func.__name__])
+
+
+def unpatch(module):
+    for module, func in wandb.patched[module.__name__]:
+        setattr(module, func, getattr(module, f"orig_{func}"))
+    wandb.patched[module.__name__] = []
+
+
+def patch_kfp():
+    patch(create_component_from_func, wandb.util.get_module("kfp.components"))
+    patch(
+        _get_function_source_definition,
+        wandb.util.get_module("kfp.components._python_op"),
+    )
+    patch(strip_type_hints, wandb.util.get_module("kfp.components._python_op"))
+
+
+def _get_function_source_definition(func: Callable) -> str:
+    func_code = inspect.getsource(func)
+
+    # Function might be defined in some indented scope (e.g. in another
+    # function). We need to handle this and properly dedent the function source
+    # code
+    func_code = textwrap.dedent(func_code)
+    func_code_lines = func_code.split("\n")
+
+    # For wandb, allow decorators (so we can use the @wandb_log decorator)
+    func_code_lines = itertools.dropwhile(
+        lambda x: not (x.startswith("def") or x.startswith("@wandb_log")),
+        func_code_lines,
+    )
+
+    if not func_code_lines:
+        raise ValueError(
+            'Failed to dedent and clean up the source of function "{}". '
+            "It is probably not properly indented.".format(func.__name__)
+        )
+
+    return "\n".join(func_code_lines)
+
+
+def create_component_from_func(
```

Comment:
I'm assuming this is copied from the kfp repo, let's add a comment linking to the original source.